File: packages/client/src/idempotent-producer.ts
/**
 * IdempotentProducer - Fire-and-forget producer with exactly-once write semantics.
 *
 * Implements Kafka-style idempotent producer pattern with:
 * - Client-provided producer IDs (zero RTT overhead)
 * - Client-declared epochs, server-validated fencing
 * - Per-batch sequence numbers for deduplication
 * - Automatic batching and pipelining for throughput
 */

import fastq from "fastq"

import { DurableStreamError, FetchError } from "./error"
import {
  PRODUCER_EPOCH_HEADER,
  PRODUCER_EXPECTED_SEQ_HEADER,
  PRODUCER_ID_HEADER,
  PRODUCER_RECEIVED_SEQ_HEADER,
  PRODUCER_SEQ_HEADER,
  STREAM_OFFSET_HEADER,
} from "./constants"
import type { queueAsPromised } from "fastq"
import type { DurableStream } from "./stream"
import type { IdempotentProducerOptions, Offset } from "./types"

/**
 * Error thrown when a producer's epoch is stale (zombie fencing).
 */
export class StaleEpochError extends Error {
  /**
   * The current epoch on the server.
   */
  readonly currentEpoch: number

  constructor(currentEpoch: number) {
    super(
      `Producer epoch is stale. Current server epoch: ${currentEpoch}. ` +
        `Call restart() or create a new producer with a higher epoch.`
    )
    this.name = `StaleEpochError`
    this.currentEpoch = currentEpoch
  }
}

/**
 * Error thrown when an unrecoverable sequence gap is detected.
 *
 * With maxInFlight > 1, HTTP requests can arrive out of order at the server,
 * causing temporary 409 responses. The client automatically handles these
 * by waiting for earlier sequences to complete, then retrying.
 *
 * This error is only thrown when the gap cannot be resolved (e.g., the
 * expected sequence is >= our sequence, indicating a true protocol violation).
 */
export class SequenceGapError extends Error {
  readonly expectedSeq: number
  readonly receivedSeq: number

  constructor(expectedSeq: number, receivedSeq: number) {
    super(
      `Producer sequence gap: expected ${expectedSeq}, received ${receivedSeq}`
    )
    this.name = `SequenceGapError`
    this.expectedSeq = expectedSeq
    this.receivedSeq = receivedSeq
  }
}

/**
 * Normalize content-type by extracting the media type (before any semicolon).
 */
function normalizeContentType(contentType: string | undefined): string {
  if (!contentType) return ``
  return contentType.split(`;`)[0]!.trim().toLowerCase()
}

/**
 * Internal type for pending batch entries.
 * Stores original data for proper JSON batching.
 */
interface PendingEntry {
  /** Original data - parsed for JSON mode batching */
  data: unknown
  /** Encoded bytes for byte-stream mode */
  body: Uint8Array
}

/**
 * Internal type for batch tasks submitted to the queue.
 */
interface BatchTask {
  batch: Array<PendingEntry>
  seq: number
}

/**
 * An idempotent producer for exactly-once writes to a durable stream.
 *
 * Features:
 * - Fire-and-forget: append() returns immediately, batches in background
 * - Exactly-once: server deduplicates using (producerId, epoch, seq)
 * - Batching: multiple appends batched into single HTTP request
 * - Pipelining: up to maxInFlight concurrent batches
 * - Zombie fencing: stale producers rejected via epoch validation
 *
 * @example
 * ```typescript
 * const stream = new DurableStream({ url: "https://..." });
 * const producer = new IdempotentProducer(stream, "order-service-1", {
 *   epoch: 0,
 *   autoClaim: true,
 * });
 *
 * // Fire-and-forget writes (synchronous, returns immediately)
 * producer.append("message 1");
 * producer.append("message 2");
 *
 * // Ensure all messages are delivered before shutdown
 * await producer.flush();
 * await producer.close();
 * ```
 */
export class IdempotentProducer {
  readonly #stream: DurableStream
  readonly #producerId: string
  #epoch: number
  #nextSeq = 0
  readonly #autoClaim: boolean
  readonly #maxBatchBytes: number
  readonly #lingerMs: number
  readonly #fetchClient: typeof fetch
  readonly #signal?: AbortSignal
  readonly #onError?: (error: Error) => void

  // Batching state
  #pendingBatch: Array<PendingEntry> = []
  #batchBytes = 0
  #lingerTimeout: ReturnType<typeof setTimeout> | null = null

  // Pipelining via fastq
  readonly #queue: queueAsPromised<BatchTask>
  readonly #maxInFlight: number
  #closed = false

  // When autoClaim is true, we must wait for the first batch to complete
  // before allowing pipelining (to know what epoch was claimed)
  #epochClaimed: boolean

  // Track sequence completions for 409 retry coordination
  // When HTTP requests arrive out of order, we get 409 errors.
  // Maps epoch -> (seq -> { resolved, error?, waiters })
  #seqState: Map<
    number,
    Map<
      number,
      {
        resolved: boolean
        error?: Error
        waiters: Array<(err?: Error) => void>
      }
    >
  > = new Map()

  /**
   * Create an idempotent producer for a stream.
   *
   * @param stream - The DurableStream to write to
   * @param producerId - Stable identifier for this producer (e.g., "order-service-1")
   * @param opts - Producer options
   */
  constructor(
    stream: DurableStream,
    producerId: string,
    opts?: IdempotentProducerOptions
  ) {
    // Validate inputs
    const epoch = opts?.epoch ?? 0
    const maxBatchBytes = opts?.maxBatchBytes ?? 1024 * 1024 // 1MB
    const maxInFlight = opts?.maxInFlight ?? 5
    const lingerMs = opts?.lingerMs ?? 5

    if (epoch < 0) {
      throw new Error(`epoch must be >= 0`)
    }
    if (maxBatchBytes <= 0) {
      throw new Error(`maxBatchBytes must be > 0`)
    }
    if (maxInFlight <= 0) {
      throw new Error(`maxInFlight must be > 0`)
    }
    if (lingerMs < 0) {
      throw new Error(`lingerMs must be >= 0`)
    }

    this.#stream = stream
    this.#producerId = producerId
    this.#epoch = epoch
    this.#autoClaim = opts?.autoClaim ?? false
    this.#maxBatchBytes = maxBatchBytes
    this.#lingerMs = lingerMs
    this.#signal = opts?.signal
    this.#onError = opts?.onError
    this.#fetchClient =
      opts?.fetch ?? ((...args: Parameters<typeof fetch>) => fetch(...args))

    this.#maxInFlight = maxInFlight

    // When autoClaim is true, epoch is not yet known until first batch completes
    // We block pipelining until then to avoid racing with the claim
    this.#epochClaimed = !this.#autoClaim

    // Initialize fastq with maxInFlight concurrency
    this.#queue = fastq.promise(this.#batchWorker.bind(this), this.#maxInFlight)

    // Handle signal abort (use { once: true } to auto-cleanup)
    if (this.#signal) {
      this.#signal.addEventListener(
        `abort`,
        () => {
          this.#rejectPendingBatch(
            new DurableStreamError(
              `Producer aborted`,
              `ALREADY_CLOSED`,
              undefined,
              undefined
            )
          )
        },
        { once: true }
      )
    }
  }

  /**
   * Append data to the stream.
   *
   * This is fire-and-forget: returns immediately after adding to the batch.
   * The message is batched and sent when:
   * - maxBatchBytes is reached
   * - lingerMs elapses
   * - flush() is called
   *
   * Errors are reported via onError callback if configured. Use flush() to
   * wait for all pending messages to be sent.
   *
   * For JSON streams, pass native objects (which will be serialized internally).
   * For byte streams, pass string or Uint8Array.
   *
   * @param body - Data to append (object for JSON streams, string or Uint8Array for byte streams)
   */
  append(body: Uint8Array | string | unknown): void {
    if (this.#closed) {
      throw new DurableStreamError(
        `Producer is closed`,
        `ALREADY_CLOSED`,
        undefined,
        undefined
      )
    }

    const isJson =
      normalizeContentType(this.#stream.contentType) === `application/json`

    let bytes: Uint8Array
    let data: unknown

    if (isJson) {
      // For JSON streams: accept native objects, serialize internally
      const json = JSON.stringify(body)
      bytes = new TextEncoder().encode(json)
      data = body
    } else {
      // For byte streams, require string or Uint8Array
      if (typeof body === `string`) {
        bytes = new TextEncoder().encode(body)
      } else if (body instanceof Uint8Array) {
        bytes = body
      } else {
        throw new DurableStreamError(
          `Non-JSON streams require string or Uint8Array`,
          `BAD_REQUEST`,
          400,
          undefined
        )
      }
      data = bytes
    }

    this.#pendingBatch.push({ data, body: bytes })
    this.#batchBytes += bytes.length

    // Check if batch should be sent immediately
    if (this.#batchBytes >= this.#maxBatchBytes) {
      this.#enqueuePendingBatch()
    } else if (!this.#lingerTimeout) {
      // Start linger timer
      this.#lingerTimeout = setTimeout(() => {
        this.#lingerTimeout = null
        if (this.#pendingBatch.length > 0) {
          this.#enqueuePendingBatch()
        }
      }, this.#lingerMs)
    }
  }

  /**
   * Send any pending batch immediately and wait for all in-flight batches.
   *
   * Call this before shutdown to ensure all messages are delivered.
   */
  async flush(): Promise<void> {
    // Clear linger timeout
    if (this.#lingerTimeout) {
      clearTimeout(this.#lingerTimeout)
      this.#lingerTimeout = null
    }

    // Enqueue any pending batch
    if (this.#pendingBatch.length > 0) {
      this.#enqueuePendingBatch()
    }

    // Wait for queue to drain
    await this.#queue.drained()
  }

  /**
   * Flush pending messages and close the producer.
   *
   * After calling close(), further append() calls will throw.
   */
  async close(): Promise<void> {
    if (this.#closed) return

    this.#closed = true

    try {
      await this.flush()
    } catch {
      // Ignore errors during close
    }
  }

  /**
   * Increment epoch and reset sequence.
   *
   * Call this when restarting the producer to establish a new session.
   * Flushes any pending messages first.
   */
  async restart(): Promise<void> {
    await this.flush()
    this.#epoch++
    this.#nextSeq = 0
  }

  /**
   * Current epoch for this producer.
   */
  get epoch(): number {
    return this.#epoch
  }

  /**
   * Next sequence number to be assigned.
   */
  get nextSeq(): number {
    return this.#nextSeq
  }

  /**
   * Number of messages in the current pending batch.
   */
  get pendingCount(): number {
    return this.#pendingBatch.length
  }

  /**
   * Number of batches currently in flight.
   */
  get inFlightCount(): number {
    return this.#queue.length()
  }

  // ============================================================================
  // Private implementation
  // ============================================================================

  /**
   * Enqueue the current pending batch for processing.
   */
  #enqueuePendingBatch(): void {
    if (this.#pendingBatch.length === 0) return

    // Take the current batch
    const batch = this.#pendingBatch
    const seq = this.#nextSeq

    this.#pendingBatch = []
    this.#batchBytes = 0
    this.#nextSeq++

    // When autoClaim is enabled and epoch hasn't been claimed yet,
    // we must wait for any in-flight batch to complete before sending more.
    // This ensures the first batch claims the epoch before pipelining begins.
    if (this.#autoClaim && !this.#epochClaimed && this.#queue.length() > 0) {
      // Wait for queue to drain, then push
      this.#queue.drained().then(() => {
        this.#queue.push({ batch, seq }).catch(() => {
          // Error handling is done in #batchWorker
        })
      })
    } else {
      // Push to fastq - it handles concurrency automatically
      this.#queue.push({ batch, seq }).catch(() => {
        // Error handling is done in #batchWorker
      })
    }
  }

  /**
   * Batch worker - processes batches via fastq.
   */
  async #batchWorker(task: BatchTask): Promise<void> {
    const { batch, seq } = task
    const epoch = this.#epoch

    try {
      await this.#doSendBatch(batch, seq, epoch)

      // Mark epoch as claimed after first successful batch
      // This enables full pipelining for subsequent batches
      if (!this.#epochClaimed) {
        this.#epochClaimed = true
      }

      // Signal success for this sequence (for 409 retry coordination)
      this.#signalSeqComplete(epoch, seq, undefined)
    } catch (error) {
      // Signal failure so waiting batches can fail too
      this.#signalSeqComplete(epoch, seq, error as Error)

      // Call onError callback if configured
      if (this.#onError) {
        this.#onError(error as Error)
      }
      throw error
    }
  }

  /**
   * Signal that a sequence has completed (success or failure).
   */
  #signalSeqComplete(
    epoch: number,
    seq: number,
    error: Error | undefined
  ): void {
    let epochMap = this.#seqState.get(epoch)
    if (!epochMap) {
      epochMap = new Map()
      this.#seqState.set(epoch, epochMap)
    }

    const state = epochMap.get(seq)
    if (state) {
      // Mark resolved and notify all waiters
      state.resolved = true
      state.error = error
      for (const waiter of state.waiters) {
        waiter(error)
      }
      state.waiters = []
    } else {
      // No waiters yet, just mark as resolved
      epochMap.set(seq, { resolved: true, error, waiters: [] })
    }

    // Clean up old entries to prevent unbounded memory growth.
    // We keep entries for the last maxInFlight * 3 sequences to handle
    // potential late 409 retries from pipelining.
    const cleanupThreshold = seq - this.#maxInFlight * 3
    if (cleanupThreshold > 0) {
      for (const oldSeq of epochMap.keys()) {
        if (oldSeq < cleanupThreshold) {
          epochMap.delete(oldSeq)
        }
      }
    }
  }

  /**
   * Wait for a specific sequence to complete.
   * Returns immediately if already completed.
   * Throws if the sequence failed.
   */
  #waitForSeq(epoch: number, seq: number): Promise<void> {
    let epochMap = this.#seqState.get(epoch)
    if (!epochMap) {
      epochMap = new Map()
      this.#seqState.set(epoch, epochMap)
    }

    const state = epochMap.get(seq)
    if (state?.resolved) {
      // Already completed
      if (state.error) {
        return Promise.reject(state.error)
      }
      return Promise.resolve()
    }

    // Not yet completed, add a waiter
    return new Promise((resolve, reject) => {
      const waiter = (err?: Error) => {
        if (err) reject(err)
        else resolve()
      }
      if (state) {
        state.waiters.push(waiter)
      } else {
        epochMap.set(seq, { resolved: false, waiters: [waiter] })
      }
    })
  }

  /**
   * Actually send the batch to the server.
   * Handles auto-claim retry on 403 (stale epoch) if autoClaim is enabled.
   * Does NOT implement general retry/backoff for network errors or 5xx responses.
   */
  async #doSendBatch(
    batch: Array<PendingEntry>,
    seq: number,
    epoch: number
  ): Promise<{ offset: Offset; duplicate: boolean }> {
    const contentType = this.#stream.contentType ?? `application/octet-stream`
    const isJson = normalizeContentType(contentType) === `application/json`

    // Build batch body based on content type
    let batchedBody: BodyInit
    if (isJson) {
      // For JSON mode: always send as array (server flattens one level)
      // Single append: [value] → server stores value
      // Multiple appends: [val1, val2] → server stores val1, val2
      const values = batch.map((e) => e.data)
      batchedBody = JSON.stringify(values)
    } else {
      // For byte mode: concatenate all chunks
      const totalSize = batch.reduce((sum, e) => sum + e.body.length, 0)
      const concatenated = new Uint8Array(totalSize)
      let offset = 0
      for (const entry of batch) {
        concatenated.set(entry.body, offset)
        offset += entry.body.length
      }
      batchedBody = concatenated
    }

    // Build URL
    const url = this.#stream.url

    // Build headers
    const headers: Record<string, string> = {
      "content-type": contentType,
      [PRODUCER_ID_HEADER]: this.#producerId,
      [PRODUCER_EPOCH_HEADER]: epoch.toString(),
      [PRODUCER_SEQ_HEADER]: seq.toString(),
    }

    // Send request
    const response = await this.#fetchClient(url, {
      method: `POST`,
      headers,
      body: batchedBody,
      signal: this.#signal,
    })

    // Handle response
    if (response.status === 204) {
      // Duplicate - idempotent success
      return { offset: ``, duplicate: true }
    }

    if (response.status === 200) {
      // Success
      const resultOffset = response.headers.get(STREAM_OFFSET_HEADER) ?? ``
      return { offset: resultOffset, duplicate: false }
    }

    if (response.status === 403) {
      // Stale epoch
      const currentEpochStr = response.headers.get(PRODUCER_EPOCH_HEADER)
      const currentEpoch = currentEpochStr
        ? parseInt(currentEpochStr, 10)
        : epoch

      if (this.#autoClaim) {
        // Auto-claim: retry with epoch+1
        const newEpoch = currentEpoch + 1
        this.#epoch = newEpoch
        this.#nextSeq = 1 // This batch will use seq 0

        // Retry with new epoch, starting at seq 0
        return this.#doSendBatch(batch, 0, newEpoch)
      }

      throw new StaleEpochError(currentEpoch)
    }

    if (response.status === 409) {
      // Sequence gap - our request arrived before an earlier sequence
      const expectedSeqStr = response.headers.get(PRODUCER_EXPECTED_SEQ_HEADER)
      const expectedSeq = expectedSeqStr ? parseInt(expectedSeqStr, 10) : 0

      // If our seq is ahead of expectedSeq, wait for earlier sequences to complete then retry
      // This handles HTTP request reordering with maxInFlight > 1
      if (expectedSeq < seq) {
        // Wait for all sequences from expectedSeq to seq-1
        const waitPromises: Array<Promise<void>> = []
        for (let s = expectedSeq; s < seq; s++) {
          waitPromises.push(this.#waitForSeq(epoch, s))
        }
        await Promise.all(waitPromises)
        // Retry now that earlier sequences have completed
        return this.#doSendBatch(batch, seq, epoch)
      }

      // If expectedSeq >= seq, something is wrong (shouldn't happen) - throw error
      const receivedSeqStr = response.headers.get(PRODUCER_RECEIVED_SEQ_HEADER)
      const receivedSeq = receivedSeqStr ? parseInt(receivedSeqStr, 10) : seq
      throw new SequenceGapError(expectedSeq, receivedSeq)
    }

    if (response.status === 400) {
      // Bad request (e.g., invalid epoch/seq)
      const error = await DurableStreamError.fromResponse(response, url)
      throw error
    }

    // Other errors - use FetchError for standard handling
    const error = await FetchError.fromResponse(response, url)
    throw error
  }

  /**
   * Clear pending batch and report error.
   */
  #rejectPendingBatch(error: Error): void {
    // Call onError callback if configured
    if (this.#onError && this.#pendingBatch.length > 0) {
      this.#onError(error)
    }
    this.#pendingBatch = []
    this.#batchBytes = 0

    if (this.#lingerTimeout) {
      clearTimeout(this.#lingerTimeout)
      this.#lingerTimeout = null
    }
  }
}

File: packages/client-py/src/durable_streams/idempotent_producer.py
"""
IdempotentProducer - Fire-and-forget producer with exactly-once write semantics.

Implements Kafka-style idempotent producer pattern with:
- Client-provided producer IDs (zero RTT overhead)
- Client-declared epochs, server-validated fencing
- Per-batch sequence numbers for deduplication
- Automatic batching and pipelining for throughput
"""

from __future__ import annotations

import asyncio
import contextlib
from dataclasses import dataclass, field
from typing import Any

import httpx

from durable_streams._errors import DurableStreamError, FetchError
from durable_streams._types import Offset

# Producer header constants
PRODUCER_ID_HEADER = "Producer-Id"
PRODUCER_EPOCH_HEADER = "Producer-Epoch"
PRODUCER_SEQ_HEADER = "Producer-Seq"
PRODUCER_EXPECTED_SEQ_HEADER = "Producer-Expected-Seq"
PRODUCER_RECEIVED_SEQ_HEADER = "Producer-Received-Seq"
STREAM_NEXT_OFFSET_HEADER = "Stream-Next-Offset"


class StaleEpochError(Exception):
    """Error thrown when a producer's epoch is stale (zombie fencing)."""

    def __init__(self, current_epoch: int) -> None:
        super().__init__(
            f"Producer epoch is stale. Current server epoch: {current_epoch}. "
            f"Call restart() or create a new producer with a higher epoch."
        )
        self.current_epoch = current_epoch


class SequenceGapError(Exception):
    """
    Error thrown when a sequence gap is detected.
    This should never happen with proper client implementation.
    """

    def __init__(self, expected_seq: int, received_seq: int) -> None:
        super().__init__(
            f"Producer sequence gap: expected {expected_seq}, received {received_seq}"
        )
        self.expected_seq = expected_seq
        self.received_seq = received_seq


@dataclass
class IdempotentAppendResult:
    """Result of an idempotent append operation."""

    offset: Offset
    duplicate: bool


@dataclass
class _PendingEntry:
    """Internal type for pending batch entries."""

    body: bytes
    future: asyncio.Future[IdempotentAppendResult] = field(
        default_factory=lambda: asyncio.get_event_loop().create_future()
    )


class IdempotentProducer:
    """
    An idempotent producer for exactly-once writes to a durable stream.

    Features:
    - Fire-and-forget: append() returns immediately, batches in background
    - Exactly-once: server deduplicates using (producerId, epoch, seq)
    - Batching: multiple appends batched into single HTTP request
    - Pipelining: up to max_in_flight concurrent batches
    - Zombie fencing: stale producers rejected via epoch validation

    Example:
        >>> async with httpx.AsyncClient() as client:
        ...     producer = IdempotentProducer(
        ...         url="https://example.com/stream",
        ...         producer_id="order-service-1",
        ...         client=client,
        ...         epoch=0,
        ...         auto_claim=True,
        ...     )
        ...
        ...     # Fire-and-forget writes
        ...     await producer.append(b"message 1")
        ...     await producer.append(b"message 2")
        ...
        ...     # Ensure all messages are delivered before shutdown
        ...     await producer.flush()
        ...     await producer.close()
    """

    def __init__(
        self,
        url: str,
        producer_id: str,
        *,
        client: httpx.AsyncClient | None = None,
        epoch: int = 0,
        auto_claim: bool = False,
        max_batch_bytes: int = 1024 * 1024,  # 1MB
        linger_ms: int = 5,
        max_in_flight: int = 5,
        content_type: str = "application/octet-stream",
    ) -> None:
        """
        Create an idempotent producer for a stream.

        Args:
            url: The full URL to the stream
            producer_id: Stable identifier for this producer (e.g., "order-service-1")
            client: Optional httpx.AsyncClient for connection reuse
            epoch: Starting epoch (default 0)
            auto_claim: If True, automatically claim higher epoch on 403
            max_batch_bytes: Maximum batch size in bytes before sending
            linger_ms: Maximum time to wait before sending a batch
            max_in_flight: Maximum concurrent batches
            content_type: Content type for appends
        """
        self._url = url
        self._producer_id = producer_id
        self._epoch = epoch
        self._next_seq = 0
        self._auto_claim = auto_claim
        self._max_batch_bytes = max_batch_bytes
        self._linger_ms = linger_ms
        self._max_in_flight = max_in_flight
        self._content_type = content_type
        self._owns_client = client is None
        self._client = client or httpx.AsyncClient(timeout=30.0)

        # Batching state
        self._pending_batch: list[_PendingEntry] = []
        self._batch_bytes = 0
        self._linger_task: asyncio.Task[None] | None = None

        # Pipelining state
        self._in_flight: dict[int, asyncio.Task[None]] = {}  # seq -> task
        self._closed = False

    @property
    def epoch(self) -> int:
        """Current epoch for this producer."""
        return self._epoch

    @property
    def next_seq(self) -> int:
        """Next sequence number to be assigned."""
        return self._next_seq

    @property
    def pending_count(self) -> int:
        """Number of messages in the current pending batch."""
        return len(self._pending_batch)

    @property
    def in_flight_count(self) -> int:
        """Number of batches currently in flight."""
        return len(self._in_flight)

    async def append(self, body: bytes | str) -> IdempotentAppendResult:
        """
        Append data to the stream.

        The message is added to the current batch and sent when:
        - max_batch_bytes is reached
        - linger_ms elapses
        - flush() is called

        Args:
            body: Data to append (bytes or str)

        Returns:
            Result with offset and duplicate flag

        Raises:
            DurableStreamError: If producer is closed
            StaleEpochError: If epoch is stale and auto_claim is False
            SequenceGapError: If sequence gap detected (should never happen)
        """
        if self._closed:
            raise DurableStreamError(
                "Producer is closed",
                code="ALREADY_CLOSED",
                status=None,
                url=self._url,
            )

        data = body.encode("utf-8") if isinstance(body, str) else body

        # Create future for this entry
        loop = asyncio.get_event_loop()
        entry = _PendingEntry(body=data, future=loop.create_future())
        self._pending_batch.append(entry)
        self._batch_bytes += len(data)

        # Check if batch should be sent immediately
        if self._batch_bytes >= self._max_batch_bytes:
            self._send_current_batch()
        elif self._linger_task is None:
            # Start linger timer
            self._linger_task = asyncio.create_task(self._linger_timeout())

        return await entry.future

    async def _linger_timeout(self) -> None:
        """Wait for linger_ms then send pending batch."""
        await asyncio.sleep(self._linger_ms / 1000.0)
        self._linger_task = None
        if self._pending_batch:
            self._send_current_batch()

    async def flush(self) -> None:
        """
        Send any pending batch immediately and wait for all in-flight batches.

        Call this before shutdown to ensure all messages are delivered.
        """
        # Cancel linger timeout
        if self._linger_task is not None:
            self._linger_task.cancel()
            with contextlib.suppress(asyncio.CancelledError):
                await self._linger_task
            self._linger_task = None

        # Loop until both pending and in-flight are drained
        while self._pending_batch or self._in_flight:
            # Try to send pending batch
            if self._pending_batch:
                self._send_current_batch()

            # If still have pending but at capacity, wait for one to complete
            if self._pending_batch and len(self._in_flight) >= self._max_in_flight:
                if self._in_flight:
                    done, _ = await asyncio.wait(
                        self._in_flight.values(), return_when=asyncio.FIRST_COMPLETED
                    )
                    # Clean up completed tasks
                    for task in done:
                        for seq, t in list(self._in_flight.items()):
                            if t is task:
                                del self._in_flight[seq]
                                break
                continue

            # Wait for all current in-flight to complete
            if self._in_flight:
                await asyncio.gather(*self._in_flight.values(), return_exceptions=True)
                self._in_flight.clear()

    async def close(self) -> None:
        """
        Flush pending messages and close the producer.

        After calling close(), further append() calls will throw.
        """
        if self._closed:
            return

        self._closed = True

        with contextlib.suppress(Exception):
            await self.flush()

        if self._owns_client:
            await self._client.aclose()

    async def restart(self) -> None:
        """
        Increment epoch and reset sequence.

        Call this when restarting the producer to establish a new session.
        Flushes any pending messages first.
        """
        await self.flush()
        self._epoch += 1
        self._next_seq = 0

    def _send_current_batch(self) -> None:
        """Send the current batch and track it in flight."""
        if not self._pending_batch:
            return

        # Wait if we've hit the in-flight limit
        if len(self._in_flight) >= self._max_in_flight:
            return

        # Take the current batch
        batch = self._pending_batch
        seq = self._next_seq

        self._pending_batch = []
        self._batch_bytes = 0
        self._next_seq += 1

        # Track this batch in flight
        task = asyncio.create_task(self._send_batch(batch, seq))
        self._in_flight[seq] = task

        # Clean up when done and maybe send pending batch
        def on_done(_task: asyncio.Task[None]) -> None:
            self._in_flight.pop(seq, None)
            # Try to send pending batch if any
            if self._pending_batch and len(self._in_flight) < self._max_in_flight:
                self._send_current_batch()

        task.add_done_callback(on_done)

    async def _send_batch(self, batch: list[_PendingEntry], seq: int) -> None:
        """Send a batch to the server."""
        try:
            result = await self._do_send_batch(batch, seq, self._epoch)

            # Resolve all entries in the batch
            for entry in batch:
                if not entry.future.done():
                    entry.future.set_result(result)
        except Exception as e:
            # Reject all entries in the batch
            for entry in batch:
                if not entry.future.done():
                    entry.future.set_exception(e)
            raise

    async def _do_send_batch(
        self, batch: list[_PendingEntry], seq: int, epoch: int
    ) -> IdempotentAppendResult:
        """
        Actually send the batch to the server.
        Handles auto-claim retry on 403 (stale epoch) if auto_claim is enabled.
        Does NOT implement general retry/backoff for network errors or 5xx responses.
        """
        # Concatenate all bodies
        concatenated = b"".join(entry.body for entry in batch)

        # Build headers
        headers = {
            "content-type": self._content_type,
            PRODUCER_ID_HEADER: self._producer_id,
            PRODUCER_EPOCH_HEADER: str(epoch),
            PRODUCER_SEQ_HEADER: str(seq),
        }

        # Send request
        response = await self._client.post(
            self._url,
            content=concatenated,
            headers=headers,
        )

        # Handle response
        if response.status_code == 204:
            # Duplicate - idempotent success
            return IdempotentAppendResult(offset="", duplicate=True)

        if response.status_code == 200:
            # Success
            result_offset = response.headers.get(STREAM_NEXT_OFFSET_HEADER, "")
            return IdempotentAppendResult(offset=result_offset, duplicate=False)

        if response.status_code == 403:
            # Stale epoch
            current_epoch_str = response.headers.get(PRODUCER_EPOCH_HEADER)
            current_epoch = int(current_epoch_str) if current_epoch_str else epoch

            if self._auto_claim:
                # Auto-claim: retry with epoch+1
                new_epoch = current_epoch + 1
                self._epoch = new_epoch
                self._next_seq = 1  # This batch will use seq 0

                # Retry with new epoch, starting at seq 0
                return await self._do_send_batch(batch, 0, new_epoch)

            raise StaleEpochError(current_epoch)

        if response.status_code == 409:
            # Sequence gap
            expected_seq_str = response.headers.get(PRODUCER_EXPECTED_SEQ_HEADER)
            received_seq_str = response.headers.get(PRODUCER_RECEIVED_SEQ_HEADER)
            expected_seq = int(expected_seq_str) if expected_seq_str else 0
            received_seq = int(received_seq_str) if received_seq_str else seq

            raise SequenceGapError(expected_seq, received_seq)

        if response.status_code == 400:
            # Bad request (e.g., invalid epoch/seq)
            text = response.text
            raise DurableStreamError(
                text or "Bad request",
                code="BAD_REQUEST",
                status=400,
                url=self._url,
            )

        # Other errors
        raise FetchError(
            f"Unexpected status {response.status_code}",
            status=response.status_code,
            url=self._url,
        )

    async def __aenter__(self) -> IdempotentProducer:
        """Async context manager entry."""
        return self

    async def __aexit__(
        self,
        exc_type: type[BaseException] | None,
        exc_val: BaseException | None,
        exc_tb: Any,
    ) -> None:
        """Async context manager exit."""
        await self.close()
